projects/Mallard/src/helpers/state.tsx

```diff
@@ -0,0 +1,39 @@
+import React, { createContext, useContext, useState, useEffect } from 'react'
+import AsyncStorage from '@react-native-community/async-storage'
+
+interface Settings {
+    apiUrl: string
+}
+const defaultSettings: Settings = {
+    apiUrl: 'https://editions-api.gutools.co.uk',
+}
+
+type StateContext = [Settings, (setting: keyof Settings, value: string) => void]
+
+const useSettings = (): StateContext => {
+    const [state, setState] = useState(defaultSettings)
+    const setSetting = (setting: keyof Settings, value: string) => {
+        setState({ [setting]: value })
+        AsyncStorage.setItem('@setting-' + setting, value)
+    }
+    useEffect(() => {
+        for (let setting of Object.keys(state)) {
+            AsyncStorage.getItem('@setting-' + setting).then(value => {
+                if (state.hasOwnProperty(setting)) {
+                    //@ts-ignore
+                    setState({ [setting]: value })
+                }
+            })
+        }
+    }, [])
+    return [state, setSetting]
+}
+
+//@ts-ignore
+export const StateContext = createContext<StateContext>(null)
```

Comment:
The error goes away with 
```ts
export const StateContext = createContext<StateContext>([
    defaultSettings,
    (..._) => {
        throw new Error('Context used without context provider')
    },
])
```

I'm not certain why the Context and Provider have to be separate. My gut instinct is
```ts
export const StateContext = createContext<StateContext>(useSettings())
```

Reply:
```
export const StateContext = createContext<StateContext>(useSettings())
```

can't use hooks outside of components!